linkerd/app/outbound/src/http/peer_proxy_errors.rs

```diff
@@ -0,0 +1,186 @@
+                    close.close();
+                }
+            }
+            Ok(response)
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        http::{Endpoint, NewServeHttp, Request, Response, StatusCode, Version},
+        test_util::{
+            self, future,
+            support::{self, connect::Connect, http_util},
+        },
+        transport::addrs::{Remote, ServerAddr},
+        Config, Outbound,
+    };
+    use hyper::{client::conn::Builder, server::conn::Http, service, Body};
+    use linkerd_app_core::{
+        errors::L5D_PROXY_ERROR,
+        io,
+        proxy::api_resolve::Metadata,
+        svc::NewService,
+        svc::{BoxNewService, BoxService},
+        tls::{ConditionalClientTls, NoClientTls},
+        Error, Infallible, ProxyRuntime,
+    };
+    use linkerd_proxy_http::BoxRequest;
+    use linkerd_tracing::test;
+    use std::net::SocketAddr;
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn connection_closes_after_response_header() {
+        let _trace = test::trace_init();
+
+        // Build the outbound server that responds with the l5d-proxy-error
+        // header set.
+        let (rt, _shutdown) = test_util::runtime();
+        let addr = SocketAddr::new([127, 0, 0, 1].into(), 12345);
+        let connect = support::connect().endpoint_fn_boxed(addr, |_: Endpoint| serve());
+        let config = test_util::default_config();
+        let mut outbound = build_outbound(config, rt, connect);
+
+        // Build the otubound service.
+        let service = outbound.new_service(Endpoint {
+            addr: Remote(ServerAddr(addr)),
+            tls: ConditionalClientTls::None(NoClientTls::Disabled),
+            metadata: Metadata::default(),
+            logical_addr: None,
+            protocol: Version::Http1,
+            opaque_protocol: false,
+        });
+
+        // Build the client that should be closed after receiving a response
+        // with the l5d-proxy-error header.
+        let mut builder = Builder::new();
+        let (mut client, bg) = http_util::connect_and_accept(&mut builder, service).await;
+
+        let mut request = Request::builder()
+            .uri("http://foo.example.com")
+            .body(Body::default())
+            .unwrap();
+        request = test_util::add_client_handle(request, addr);
+        let response = http_util::http_request(&mut client, request).await.unwrap();
+        assert_eq!(response.status(), StatusCode::BAD_GATEWAY);
+        let message = response
+            .headers()
+            .get(L5D_PROXY_ERROR)
+            .expect("response did not contain l5d-proxy-error header");
+        assert_eq!(message, "proxy received invalid response");
+
+        // The client's background task should be completed without dropping
+        // the client because the connection was closed after encountering a
+        // response that contains the l5d-proxy-error header.
+        let _ = bg.await;
```

Comment:
I don't understand what this is actually testing or how it's wired up with the client handle.

Specifically https://github.com/linkerd/linkerd2-proxy/pull/1183 changes the client handle test utility to return the close future; and the close future never fires? Why is that? How does the background task get notified about the client handle signaling a close?

Reply:
Or, presumably `NewServeHttp` adds a client handle? If so, why do we have to call `add_client_handle` in the test?

Reply:
It may be better to just narrow this test to only test the behavior in this module without actually initiating TCP clients and servers, and the whole stack. i.e. just test that close future fires.

Reply:
https://github.com/linkerd/linkerd2-proxy/pull/1183/commits/7e990193c81ecd0417f8e1c8bc0fb95722988ae5